### PR TITLE
bug fix in learn, primary seond a prepared mutation with old ballot, but learner has newer ballot

### LINKED ISSUE #496: smaller ballot mutation prepare crash during learning
Given three replica servers, R1, R2 and R3. Consider this:
- Only two replicas in a replica group, R1 the primary, R2 the secondary. And the (ballot, prepare_decree, commit_decree) of R1 is (1, 0, 0). similarly, R2(1, 0, 0)
- R1, the primary, got a client request, finally the status changed to: R1(1, 1, 1), R2(1, 1, 0)
- R1 send a group check, then status changed to: R1(1, 1, 1), R2(1, 1, 1)
- Then R3 add to the replica group, so the status of R1 and R2 changed to: R1(2, 1, 1), R2(2, 1, 1)
- Then R1 crashed, then restart very soon, in which case the primary is not switched. The status will changed to: R1(2, 1, 0), R2(2, 1, 1). Notice the R1(2, 1, 0) is desirable because the commit decree is only a memory state.
- Then R1 starts to replay the prepare_list, so a new mutation(**ballot:2, decree:1**) is re-appended to **the prepare log of R1** and send to R2. For R2, mutation(ballot:2, decree:1) will be replied immediately, because it is already **commited**. So, at last, the status is: R1(2, 1, 1), R2(2, 1, 1) again. But for mutation with decree 1, **it will have a ballot 2 in R1 but ballot 1 in R2.**
- Then R1 crashed again. It is not lucky this time. R2 becomes the primary, and the recovered R1 turned a learn. Currently the status is: R2(2, 1, 1), R1(2, 1, 0).  Learner R1 will have to learn mutation(decree:1) from R2. But R1 have a larger ballot for decree 1, so the assertion failed.

Ways to fix this:
(1) stop to send the last_commited_decree in group_check, in which case the **commit invariant** in PacificA will be kept. 
(2) A secondary will prepare a mutation as long as a larger ballot is met, regardless of the commit decree.
(3) When replaying prepare_list, the ballot of a mutation should never be changed. Additionally, primary will need to send its **current ballot**, which should be no smaller than the mutation's ballot.  

Fix (1) and (2) is a tiny fix, but both abandon some current optimization. I'd prefer (3), but I don't know if other problems are introduced. 

What do you say? @qinzuoyan @imzhenyu

### PR DESCRIPTION
**Pull request by shengofsun**
resolves #496

### CODE REVIEW

File: src/dist/replication/lib/replica_learn.cpp [hunk 1]
```diff
@@ -592,22 +592,31 @@ void replica::on_learn_reply(
...
-
-                // then we prepare
-                _prepare_list->prepare(mu, partition_status::PS_POTENTIAL_SECONDARY);
-
-                if (cache_range.first == 0 || mu->data.header.decree < cache_range.first)
-                    cache_range.first = mu->data.header.decree;
-                if (cache_range.second == 0 || mu->data.header.decree > cache_range.second)
-                    cache_range.second = mu->data.header.decree;
+                mutation_ptr existing_mutation = _prepare_list->get_mutation_by_decree(mu->data.header.decree);
+                if (existing_mutation!=nullptr && existing_mutation->data.header.ballot >= mu->data.header.ballot)
```
Comment (imzhenyu) [inline src/dist/replication/lib/replica_learn.cpp]: the filter should be applied to prepare list only - the logs may be truncated already so logging is always necessary in this case.

Comment (shengofsun) [inline src/dist/replication/lib/replica_learn.cpp]: Is it convinient for you if I call you up？

Comment (imzhenyu) [inline src/dist/replication/lib/replica_learn.cpp]: sure

**Comment by shengofsun** [comment]
@imzhenyu fix this as discussed

**Pull request merged**